controllers/update_status.go

```diff
@@ -624,6 +640,75 @@ func validateProcessGroup(ctx context.Context, r *FoundationDBClusterReconciler,
+					if *taintConfiguredKey.DurationInSeconds < 0 {
+						logger.Info("TaintReplacementOption is disabled", "Key",
+							taintConfiguredKey.Key, "DurationInSeconds", taintConfiguredKey.DurationInSeconds)
+						continue
+					}
+					if taint.Key == *taintConfiguredKey.Key || *taintConfiguredKey.Key == "*" {
+						hasValidTaint = true
+						logger.Info("MX Debug Info UpdateTaint", "Pod", pod.Name, "Add Condition", "NodeTaintDetected")
+						processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintDetected, true, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+						// Use node taint's timestamp as NodeTaintDetected condition's starting time
+						if taint.TimeAdded != nil && taint.TimeAdded.Time.Unix() < *processGroupStatus.GetConditionTime(fdbv1beta2.NodeTaintDetected) {
+							logger.Info("MX Debug Info Update NodeTaintDetected Time", "Pod", pod.Name,
+								"OriginalTime", processGroupStatus.GetConditionTime(fdbv1beta2.NodeTaintDetected),
+								"NewTime", taint.TimeAdded.Unix())
+							processGroupStatus.UpdateConditionTime(fdbv1beta2.NodeTaintDetected, taint.TimeAdded.Unix())
+						}
+						for _, processGroupCondition := range processGroupStatus.ProcessGroupConditions {
+							logger.Info("MX Debug Info Check to add NodeTaintReplacing Condition", "Pod", pod.Name,
+								"Now", time.Now().Unix(), "ConditionTimestamp", processGroupCondition.Timestamp,
+								"TaintConfiguredDuration", *taintConfiguredKey.DurationInSeconds)
+							if processGroupCondition.ProcessGroupConditionType == fdbv1beta2.NodeTaintDetected &&
+								time.Now().Unix()-processGroupCondition.Timestamp > *taintConfiguredKey.DurationInSeconds {
+								processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintReplacing, true, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+								logger.Info("Add NodeTaintReplacing condition", "Pod", pod.Name, "Node", node.Name,
+									"TaintKey", taint.Key, "TaintDetectedTime", processGroupCondition.Timestamp,
+									"TaintDuration", int64(time.Since(time.Unix(processGroupCondition.Timestamp, 0))),
+									"TaintValue", taint.Value, "TaintEffect", taint.Effect,
+									"NodeTaintDetectedConditionTimestamp", time.Unix(processGroupCondition.Timestamp, 0),
+									"Duration", time.Duration(*taintConfiguredKey.DurationInSeconds),
+									"Now", time.Now(),
+									"NodeTaintReplacingCondition", processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected))
+							}
+						}
+					}
+				}
+			}
+			if !hasValidTaint {
+				// Remove NodeTaintDetected condition if the pod is no longer on a tainted node;
+				// This is needed especially when a node's status is flapping
+				// We do not remove NodeTaintDetected because the NodeTaintDetected condition is only added when taint is there for configured long time
+				processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintDetected, false, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+			}
+		}
```

Comment:
It probably make sense to move this logic into a small method (this will make the testing easier and makes reading the method easier. The proposed design: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/docs/design/node_taint_eviction.md#proposed-design proposed to query all nodes in a single call, if we are not doing this, we should update the design later to reflect the implementation, independent of the documentation for the new feature.

Reply:
will update the design doc.